airbyte_cdk/sources/declarative/stream_slicers/declarative_partition_generator.py

```diff
@@ -46,17 +57,24 @@ def __init__(
         json_schema: Mapping[str, Any],
         retriever: Retriever,
         message_repository: MessageRepository,
+        max_records_limit: Optional[int],
         stream_slice: StreamSlice,
     ):
         self._stream_name = stream_name
         self._json_schema = json_schema
         self._retriever = retriever
         self._message_repository = message_repository
+        self._max_records_limit = max_records_limit
         self._stream_slice = stream_slice
         self._hash = SliceHasher.hash(self._stream_name, self._stream_slice)
 
     def read(self) -> Iterable[Record]:
         for stream_data in self._retriever.read_records(self._json_schema, self._stream_slice):
```

Comment:
I'm not sure anymore how generator works but it feels like this will perform an HTTP request even if we hit the limit before processing the slice. Should we have a `if total_record_counter >= self._max_records_limit:` condition before the for loop as well?

Reply:
yep makes sense, that should proactively stop from making the first request. i'll fix and add a test for this